Installation via bdist_wheel and bdist_egg can't find resampy/interp.c

I've had a couple of issues installing resampy when trying to install librosa, with errors similar to the following:

```
Searching for resampy>=0.1.0
Reading https://pypi.python.org/simple/resampy/
Best match: resampy 0.1.0
Downloading https://pypi.python.org/packages/e1/8d/9fbc8e1dba8a83e4be47cc4a906dcb3ca5959bc7e4da1d388e64c2d20a5e/resampy-0.1.0.tar.gz#md5=3fec740b9dcc40397efd4d1b4e4ad532
Processing resampy-0.1.0.tar.gz
Writing /tmp/easy_install-_2lda2x_/resampy-0.1.0/setup.cfg
Running resampy-0.1.0/setup.py -q bdist_egg --dist-dir /tmp/easy_install-_2lda2x_/resampy-0.1.0/egg-dist-tmp-14darseq
warning: no files found matching '*.pxd' under directory 'resampy'
gcc: error: resampy/interp.c: No such file or directory
gcc: fatal error: no input files
compilation terminated.
```

In general, this can be worked around by doing `pip install resampy` which gives similar errors about missing resampy/interp.c during bdist_wheel, but then falls back on installing via setup.py install, something like (on windows):

```
>pip install resampy
Collecting resampy
  Downloading resampy-0.1.0.tar.gz (320kB)
    100% |████████████████████████████████| 327kB 101kB/s 
Requirement already satisfied (use --upgrade to upgrade): numpy>=1.10 in c:\miniconda3\envs\sct\lib\site-packages (from resampy)
Requirement already satisfied (use --upgrade to upgrade): scipy>=0.13 in c:\miniconda3\envs\sct\lib\site-packages (from resampy)
Requirement already satisfied (use --upgrade to upgrade): six>=1.3 in c:\miniconda3\envs\sct\lib\site-packages (from resampy)
Collecting Cython>=0.21 (from resampy)
  Downloading Cython-0.24-cp35-none-win_amd64.whl (1.9MB)
    100% |████████████████████████████████| 1.9MB 433kB/s 
Building wheels for collected packages: resampy
  Running setup.py bdist_wheel for resampy ... error
  Complete output from command C:\Miniconda3\envs\sct\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\michael\\AppData\\Local\\Temp\\pip-build-7wr0ovfk\\resampy\\setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" bdist_wheel -d C:\Users\michael\AppData\Local\Temp\tmpmu8ren92pip-wheel- --python-tag cp35:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.5
  creating build\lib.win-amd64-3.5\resampy
  copying resampy\core.py -> build\lib.win-amd64-3.5\resampy
  copying resampy\filters.py -> build\lib.win-amd64-3.5\resampy
  copying resampy\version.py -> build\lib.win-amd64-3.5\resampy
  copying resampy\__init__.py -> build\lib.win-amd64-3.5\resampy
  creating build\lib.win-amd64-3.5\resampy\data
  copying resampy\data\kaiser_best.npz -> build\lib.win-amd64-3.5\resampy\data
  copying resampy\data\kaiser_fast.npz -> build\lib.win-amd64-3.5\resampy\data
  running build_ext
  building 'resampy.interp' extension
  creating build\temp.win-amd64-3.5
  creating build\temp.win-amd64-3.5\Release
  creating build\temp.win-amd64-3.5\Release\resampy
  D:\Dev\VS\VC\BIN\amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -IC:\Miniconda3\envs\sct\lib\site-packages\numpy\core\include -IC:\Miniconda3\envs\sct\include -IC:\Miniconda3\envs\sct\include -ID:\Dev\VS\VC\INCLUDE -ID:\Dev\VS\VC\ATLMFC\INCLUDE "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10586.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10586.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10586.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10586.0\winrt" -ID:\Dev\VS\VC\include "-IC:\Program Files (x86)\Windows Kits\10\Include\10.0.10586.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\Include\10.0.10586.0\um" "-IC:\Program Files (x86)\Windows Kits\10\Include\10.0.10586.0\ucrt" /Tcresampy/interp.c /Fobuild\temp.win-amd64-3.5\Release\resampy/interp.obj
  interp.c
  c1: fatal error C1083: Cannot open source file: 'resampy/interp.c': No such file or directory
  error: command 'D:\\Dev\\VS\\VC\\BIN\\amd64\\cl.exe' failed with exit status 2

  ----------------------------------------
  Failed building wheel for resampy
  Running setup.py clean for resampy
Failed to build resampy
Installing collected packages: Cython, resampy
  Running setup.py install for resampy ... done
Successfully installed Cython-0.24 resampy-0.1.0
```

Let me know if you need any further info from me.

--- a/.travis.yml
+++ b/.travis.yml
@@ -35,6 +35,7 @@ before_install:
 
 install:
     - pip install python-coveralls pytest-faulthandler
+    - python setup.py build_ext -i
     - pip install -e .[docs,tests]
 
 script:

--- a/MANIFEST.in
+++ b/MANIFEST.in
@@ -1,4 +1,5 @@
 include *.md
 include *.py
 recursive-include resampy *.py *.pyx *.pxd *.npz
+recursive-include tools *.py
 include LICENSE

--- a/setup.py
+++ b/setup.py
@@ -1,25 +1,49 @@
+#!/usr/bin/env python
+'''resampy setup script'''
+
+from __future__ import print_function
+import imp
+import sys
+import os
+import subprocess
+
 from setuptools import setup
 from setuptools.extension import Extension
 
 import numpy as np
 
-import imp
+PACKAGE_NAME = 'resampy'
 
-version = imp.load_source('resampy.version', 'resampy/version.py')
+VERSION = imp.load_source('resampy.version', 'resampy/version.py')
+
+def generate_cython(package):
+    """Cythonize all sources in the package"""
+    cwd = os.path.abspath(os.path.dirname(__file__))
+    print("Cythonizing sources")
+    p = subprocess.call([sys.executable,
+                         os.path.join(cwd, 'tools', 'cythonize.py'),
+                         package],
+                        cwd=cwd)
+    if p != 0:
+        raise RuntimeError("Running cythonize failed!")
+
+
+EXTENSIONS = [Extension("resampy.interp", ["resampy/interp.c"],
+                        include_dirs=[np.get_include()])]
+
+generate_cython(PACKAGE_NAME)
 
 setup(
     author="Brian McFee",
     author_email="brian.mcfee@nyu.edu",
-    name='resampy',
-    version=version.version,
+    name=PACKAGE_NAME,
+    version=VERSION.version,
     url='https://github.com/bmcfee/resampy',
     download_url='https://github.com/bmcfee/resampy/releases',
     description='Efficient signal resampling',
     license='ISC',
-    ext_modules=[Extension("resampy.interp",
-                           ["resampy/interp.pyx"],
-                           include_dirs=[np.get_include()] )],
-    packages=['resampy'],
+    ext_modules=EXTENSIONS,
+    packages=[PACKAGE_NAME],
     package_data={'resampy': ['data/*']},
     install_requires=[
         'numpy>=1.10',
@@ -45,5 +69,7 @@ setup(
         "Programming Language :: Python :: 2",
         "Programming Language :: Python :: 2.7",
         "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.4",
+        "Programming Language :: Python :: 3.5",
     ],
-) 
+)

--- a/tools/cythonize.py
+++ b/tools/cythonize.py
@@ -1,0 +1,198 @@
+#!/usr/bin/env python
+""" cythonize
+
+Cythonize pyx files into C files as needed.
+
+Usage: cythonize [root_dir]
+
+Checks pyx files to see if they have been changed relative to their
+corresponding C files.  If they have, then runs cython on these files to
+recreate the C files.
+
+The script detects changes in the pyx/pxd files using checksums
+[or hashes] stored in a database file
+
+Simple script to invoke Cython on all .pyx
+files; while waiting for a proper build system. Uses file hashes to
+figure out if rebuild is needed.
+
+It is called by ./setup.py sdist so that sdist package can be installed without
+cython
+
+Originally written by Dag Sverre Seljebotn, and adapted from statsmodel 0.6.1
+(Modified BSD 3-clause)
+
+We copied it for scikit-learn.
+
+Note: this script does not check any of the dependent C libraries; it only
+operates on the Cython .pyx files or their corresponding Cython header (.pxd)
+files.
+"""
+# Author: Arthur Mensch <arthur.mensch@inria.fr>
+# Author: Raghav R V <rvraghav93@gmail.com>
+#
+# License: BSD 3 clause
+# see http://github.com/scikit-learn/scikit-learn
+
+
+from __future__ import division, print_function, absolute_import
+
+import os
+import re
+import sys
+import hashlib
+import subprocess
+
+HASH_FILE = 'cythonize.dat'
+
+
+# WindowsError is not defined on unix systems
+try:
+    WindowsError
+except NameError:
+    WindowsError = None
+
+
+def cythonize(cython_file, gen_file):
+    try:
+        from Cython.Compiler.Version import version as cython_version
+        from distutils.version import LooseVersion
+        if LooseVersion(cython_version) < LooseVersion('0.21'):
+            raise Exception('Building scikit-learn requires Cython >= 0.21')
+
+    except ImportError:
+        pass
+
+    flags = ['--fast-fail']
+    if gen_file.endswith('.cpp'):
+        flags += ['--cplus']
+
+    try:
+        try:
+            rc = subprocess.call(['cython'] +
+                                 flags + ["-o", gen_file, cython_file])
+            if rc != 0:
+                raise Exception('Cythonizing %s failed' % cython_file)
+        except OSError:
+            # There are ways of installing Cython that don't result in a cython
+            # executable on the path, see scipy issue gh-2397.
+            rc = subprocess.call([sys.executable, '-c',
+                                  'import sys; from Cython.Compiler.Main '
+                                  'import setuptools_main as main;'
+                                  ' sys.exit(main())'] + flags +
+                                 ["-o", gen_file, cython_file])
+            if rc != 0:
+                raise Exception('Cythonizing %s failed' % cython_file)
+    except OSError:
+        raise OSError('Cython needs to be installed')
+
+
+def load_hashes(filename):
+    """Load the hashes dict from the hashfile"""
+    # { filename : (sha1 of header if available or 'NA',
+    #               sha1 of input,
+    #               sha1 of output) }
+
+    hashes = {}
+    try:
+        with open(filename, 'r') as cython_hash_file:
+            for hash_record in cython_hash_file:
+                (filename, header_hash,
+                 cython_hash, gen_file_hash) = hash_record.split()
+                hashes[filename] = (header_hash, cython_hash, gen_file_hash)
+    except (KeyError, ValueError, AttributeError, IOError):
+        hashes = {}
+    return hashes
+
+
+def save_hashes(hashes, filename):
+    """Save the hashes dict to the hashfile"""
+    with open(filename, 'w') as cython_hash_file:
+        for key, value in hashes.items():
+            cython_hash_file.write("%s %s %s %s\n"
+                                   % (key, value[0], value[1], value[2]))
+
+
+def sha1_of_file(filename):
+    h = hashlib.sha1()
+    with open(filename, "rb") as f:
+        h.update(f.read())
+    return h.hexdigest()
+
+
+def clean_path(path):
+    """Clean the path"""
+    path = path.replace(os.sep, '/')
+    if path.startswith('./'):
+        path = path[2:]
+    return path
+
+
+def get_hash_tuple(header_path, cython_path, gen_file_path):
+    """Get the hashes from the given files"""
+
+    header_hash = (sha1_of_file(header_path)
+                   if os.path.exists(header_path) else 'NA')
+    from_hash = sha1_of_file(cython_path)
+    to_hash = (sha1_of_file(gen_file_path)
+               if os.path.exists(gen_file_path) else 'NA')
+
+    return header_hash, from_hash, to_hash
+
+
+def cythonize_if_unchanged(path, cython_file, gen_file, hashes):
+    full_cython_path = os.path.join(path, cython_file)
+    full_header_path = full_cython_path.replace('.pyx', '.pxd')
+    full_gen_file_path = os.path.join(path, gen_file)
+
+    current_hash = get_hash_tuple(full_header_path, full_cython_path,
+                                  full_gen_file_path)
+
+    if current_hash == hashes.get(clean_path(full_cython_path)):
+        print('%s has not changed' % full_cython_path)
+        return
+
+    print('Processing %s' % full_cython_path)
+    cythonize(full_cython_path, full_gen_file_path)
+
+    # changed target file, recompute hash
+    current_hash = get_hash_tuple(full_header_path, full_cython_path,
+                                  full_gen_file_path)
+
+    # Update the hashes dict with the new hash
+    hashes[clean_path(full_cython_path)] = current_hash
+
+
+def check_and_cythonize(root_dir):
+    print(root_dir)
+    hashes = load_hashes(HASH_FILE)
+
+    for cur_dir, dirs, files in os.walk(root_dir):
+        for filename in files:
+            if filename.endswith('.pyx'):
+                gen_file_ext = '.c'
+                # Cython files with libcpp imports should be compiled to cpp
+                with open(os.path.join(cur_dir, filename), 'rb') as f:
+                    data = f.read()
+                    m = re.search(b"libcpp", data, re.I | re.M)
+                    if m:
+                        gen_file_ext = ".cpp"
+                cython_file = filename
+                gen_file = filename.replace('.pyx', gen_file_ext)
+                cythonize_if_unchanged(cur_dir, cython_file, gen_file, hashes)
+
+                # Save hashes once per module. This prevents cythonizing prev.
+                # files again when debugging broken code in a single file
+                save_hashes(hashes, HASH_FILE)
+
+
+def main(root_dir):
+    check_and_cythonize(root_dir)
+
+
+if __name__ == '__main__':
+    try:
+        root_dir_arg = sys.argv[1]
+    except IndexError:
+        raise ValueError("Usage: python cythonize.py <root_dir>")
+    main(root_dir_arg)

